src/librustc_mir/interpret/intern.rs

```diff
@@ -353,7 +380,29 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
         } else if ecx.memory.dead_alloc_map.contains_key(&alloc_id) {
             // dangling pointer
             throw_unsup!(ValidationFailure("encountered dangling pointer in final constant".into()))
-        } else if ecx.tcx.alloc_map.lock().get(alloc_id).is_none() {
+        } else if let Some(_) = ecx.tcx.alloc_map.lock().get(alloc_id) {
+            // We have hit an `AllocId` that belongs to an already interned static,
+            // and are thus not interning any further.
+
+            // E.g. this should be unreachable for `InternKind::Promoted` except for allocations
+            // created for string and byte string literals, since these are interned immediately
+            // at creation time.
+
+            // FIXME(oli-obk): Since we are also checking things during interning,
+            // we should probably continue doing those checks no matter what we encounter.
```

Comment:
```suggestion
            // we could possibly continue doing those checks no matter what we encounter.
```
I am very far from convinced that that would be a reasonable thing to do. We should check the new thing we add, but I don't see why we'd re-check the old things that were already created. For example, we have to ensure that interning a `const` doesn't add anything mutable, but there is nothing fundamentally wrong with having a `const` refer to something mutable that *already existed*.

So, I think "probably" and "we basically have to" are way too strong words here. My intuition is, doing that extra check is wrong.

Reply:
> re-check the old things that were already created. 

Because the types at which we are checking may have changed. If you transmuted a pointer to an already interned allocation, you may have changed the way the data should get validated

Reply:
That's a validation thing. But the things that are checked here, like mutability, are of a different nature -- as in the example I gave.

Validation *already* follows references into other statics/promoteds if they are in the same crate.

Reply:
This is also non-critical for the regression fix so can be factored into a separate PR, right?

Reply:
definitely, that's why it's a FIXME

> That's a validation thing. But the things that are checked here, like mutability, are of a different nature -- as in the example I gave.

What I mean is if you have a relocation to a static with interior mutability behind a raw pointer, then validation won't catch this, but interning can realize that there's a mutable allocation reachable from a constant.

Reply:
FIXME still means "here's something that needs fixing", and I have strong reservations that this is the case, hence my reluctance to land this.

> What I mean is if you have a relocation to a static with interior mutability behind a raw pointer, then validation won't catch this, but interning can realize that there's a mutable allocation reachable from a constant.

I understand. And I keep saying that this doesn't matter. Interning doesn't check for mutable allocations reachable for constants elsewhere (through references) either, only validation does (and more by accident).

There is no reason, that I can see, that we'd ever want to run-time dynamically check that a constant doesn't point to a mutable static. We don't have such a check right now. What we have is a check making sure that a constant doesn't *read from* a mutable static, and through validation that happens to also exclude some constants pointing to mutable statics, but that is an accident, not by design. At least, that is my understanding of the situation.